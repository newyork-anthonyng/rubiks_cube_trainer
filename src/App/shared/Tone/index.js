import React, { Component } from "react";
import PropTypes from "prop-types";

class Tone extends Component {
  componentDidMount() {
    this.oscillator = null;
    this.frequency = 440;

    if (this.props.play) {
      this.createOscillator();
      this.oscillator.start();
    }
  }

  componentDidUpdate(prevProps) {
    if (!prevProps.play && this.props.play) {
      this.createOscillator();
      this.oscillator.start();
      this.oscillator.stop(
        this.props.audioContext.currentTime + this.props.length
      );
    }
  }

  createOscillator() {
    const { audioContext } = this.props;
    this.oscillator = audioContext.createOscillator();
    this.oscillator.connect(audioContext.destination);
    this.oscillator.frequency.value = this.frequency;
  }

  render() {
    return null;
  }
}

Tone.propTypes = {
  play: PropTypes.bool,
  length: PropTypes.number,
  audioContext: PropTypes.object
};

Tone.defaultProps = {
  length: 0.05,
  play: false
};

export default Tone;
