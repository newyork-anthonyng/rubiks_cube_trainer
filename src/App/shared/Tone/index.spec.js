import React from "react";
import { shallow, mount } from "enzyme";
import Tone from "./";

const mockOscillator = {
  start: jest.fn(),
  stop: jest.fn(),
  connect: jest.fn(),
  frequency: {
    value: undefined
  }
};
const mockAudioContext = {
  createOscillator: jest.fn(() => mockOscillator),
  destination: "mock-audio-context-destination",
  currentTime: 42
};

const defaultProps = {
  play: false,
  length: 0.42,
  audioContext: mockAudioContext
};

beforeEach(() => {
  mockAudioContext.createOscillator.mockClear();

  mockOscillator.start.mockClear();
  mockOscillator.stop.mockClear();
  mockOscillator.connect.mockClear();
});

it("should render nothing", () => {
  const wrapper = shallow(<Tone {...defaultProps} />);

  expect(wrapper.html()).toEqual(null);
});

describe("should play a sound", () => {
  it("when mounting and play=true", () => {
    const wrapper = mount(<Tone {...defaultProps} play={true} />);

    expect(mockAudioContext.createOscillator).toHaveBeenCalledTimes(1);
    expect(mockOscillator.connect).toHaveBeenCalledTimes(1);
    expect(mockOscillator.connect.mock.calls[0]).toMatchSnapshot();
    expect(mockOscillator.start).toHaveBeenCalledTimes(1);
    expect(mockOscillator.stop).toHaveBeenCalledTimes(1);
    expect(mockOscillator.stop.mock.calls[0]).toMatchSnapshot();
    expect(mockOscillator.frequency.value).toMatchSnapshot();
  });

  it("when changing from play=false to play=true", () => {
    const wrapper = mount(<Tone {...defaultProps} />);

    expect(mockOscillator.start).not.toHaveBeenCalled();

    wrapper.setProps({
      play: true
    });

    expect(mockAudioContext.createOscillator).toHaveBeenCalledTimes(1);
    expect(mockOscillator.connect).toHaveBeenCalledTimes(1);
    expect(mockOscillator.connect.mock.calls[0]).toMatchSnapshot();
    expect(mockOscillator.start).toHaveBeenCalledTimes(1);
    expect(mockOscillator.stop).toHaveBeenCalledTimes(1);
    expect(mockOscillator.stop.mock.calls[0]).toMatchSnapshot();
    expect(mockOscillator.frequency.value).toMatchSnapshot();
  });
});

it("should not play a sound when changing from play=true to play=false", () => {
  const wrapper = mount(<Tone {...defaultProps} play={true} />);

  mockAudioContext.createOscillator.mockClear();
  mockOscillator.start.mockClear();
  mockOscillator.stop.mockClear();
  mockOscillator.connect.mockClear();

  wrapper.setProps({
    play: false
  });

  expect(mockAudioContext.createOscillator).not.toHaveBeenCalled();
  expect(mockOscillator.connect).not.toHaveBeenCalled();
  expect(mockOscillator.start).not.toHaveBeenCalled();
  expect(mockOscillator.stop).not.toHaveBeenCalled();
});
