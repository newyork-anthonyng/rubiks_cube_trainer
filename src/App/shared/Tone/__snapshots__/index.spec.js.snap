// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should play a sound when changing from play=false to play=true 1`] = `
Array [
  "mock-audio-context-destination",
]
`;

exports[`should play a sound when changing from play=false to play=true 2`] = `
Array [
  42.42,
]
`;

exports[`should play a sound when changing from play=false to play=true 3`] = `440`;

exports[`should play a sound when mounting and play=true 1`] = `
Array [
  "mock-audio-context-destination",
]
`;

exports[`should play a sound when mounting and play=true 2`] = `
Array [
  42.42,
]
`;

exports[`should play a sound when mounting and play=true 3`] = `440`;
